cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -50,34 +50,27 @@ extern "C" {
 #define SetGPIO(a) do { GPOS = a; } while (0)
 #define ClearGPIO(a) do { GPOC = a; } while (0)
 
+enum class ExpiryState : uint32_t {OFF = 0, ON = 1, UPDATE = 2}; // for UPDATE, the NMI computes the exact expiry cycle and transitions to ON
+
 // Waveform generator can create tones, PWM, and servos
 typedef struct {
-  uint32_t nextServiceCycle;   // ESP cycle timer when a transition required
-  uint32_t expiryCycle;        // For time-limited waveform, the cycle when this waveform must stop
-  uint32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
-  uint32_t nextTimeLowCycles;  // Copy over high->low to keep smooth waveform
+  uint32_t nextServiceCycle;           // ESP cycle timer when a transition required
+  volatile int32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
+  volatile int32_t nextTimeLowCycles;  // Copy over high->low to keep smooth waveform
+  volatile uint32_t expiryCycle;       // For time-limited waveform, the cycle when this waveform must stop. In ExpiryState::UPDATE, temporarily holds relative cycle count
+  volatile ExpiryState hasExpiry;      // OFF: expiryCycle (temporarily) ignored. UPDATE: expiryCycle is zero-based, NMI will recompute 
```

Comment:
This add'l entry uses 17*4 = 68 bytes of heap extra.  Would it be possible to pack this into one or two uint32s (like the waveformEnabled, etc.) to bring it back down?  There's only 2 bits of state really used per pin..

Reply:
It's supposed to be atomic, and going with unit8_t forced alignment padding, so IIRC nothing is saved.